qir/qat/GroupingPass/GroupingPass.cpp

```diff
@@ -0,0 +1,262 @@
+                        instr.replaceAllUsesWith(new_instr);
+                        to_delete.push_back(&instr);
+
+                        // new_quantum_measure_cycle = true;
+                        continue;
+                    }
+
+                    if (is_quantum)
+                    {
+
+                        if (new_quantum_measure_cycle)
+                        {
+                            // Creating new measure context
+                            quantum_builder.CreateBr(readout_block);
+                            quantum_block = llvm::BasicBlock::Create(
+                                context, "quantum", tail_classical->getParent(), tail_classical);
+
+                            readout_builder.CreateBr(quantum_block);
+                            readout_block = llvm::BasicBlock::Create(
+                                context, "readout", tail_classical->getParent(), tail_classical);
+
+                            // Moving the builders to the new blocks
+                            quantum_builder.SetInsertPoint(quantum_block);
+                            readout_builder.SetInsertPoint(readout_block);
+                            new_quantum_measure_cycle = false;
+                        }
+
+                        auto new_instr = instr.clone();
+                        quantum_builder.Insert(new_instr);
+                        instr.replaceAllUsesWith(new_instr);
+                        to_delete.push_back(&instr);
+                        continue;
+                    }
+                }
+            }
+
+            // Only the terminator survives in the tail block
+            if (instr.isTerminator())
+            {
+                continue;
+            }
+
+            // Check if depends on readout
+            if (depends_on_readout.find(&instr) != depends_on_readout.end())
+            {
+                for (auto user : instr.users())
+                {
+                    depends_on_readout.insert(user);
+                }
+
+                continue;
+            }
+
+            // Moving remaining
+            auto new_instr = instr.clone();
+            classical_builder.Insert(new_instr);
+            instr.replaceAllUsesWith(new_instr);
+            to_delete.push_back(&instr);
+        }
+
+        classical_builder.CreateBr(orig_quantum);
+        quantum_builder.CreateBr(readout_block);
+        readout_builder.CreateBr(tail_classical);
+
+        for (auto it = to_delete.rbegin(); it != to_delete.rend(); ++it)
+        {
+            auto ptr = *it;
+            if (ptr->use_empty())
+            {
+                ptr->eraseFromParent();
+            }
+        }
+    }
+
+    llvm::PreservedAnalyses GroupingPass::run(llvm::Module& module, llvm::ModuleAnalysisManager& mam)
+    {
+        auto& result = mam.getResult<GroupingAnalysisPass>(module);
+
+        for (auto* block : result.qc_cc_blocks)
+        {
+            expandBlockQuantumClassical(module, block);
+        }
```

Comment:
I'm a little confused as to why this is needed in addition to the use of `expandBlockQuantumMeasureClassical`. Wouldn't any block that has quantum and classical but no measurement also be handled by `expandBlockQuantumMeasureClassical`? Using the before/after graph you created in the header file, I would've assumed that passing a block with no measurements would just result in empty "bb mc" and "bb tail cc" blocks, which collapses trivially to the specialized no measurements case. I'm unclear on what value processing blocks without measurements separately provides.

Reply:
Yes, they can indeed. The two functions just reflect progressive work and the initial thought that they need to be handled separately. One thing worth noting though is that there may be merits to separating it into two functions as the former amounts to fire-and-forget for a async function where as the latter is fire and await. Whether this is something we should have support for is a question that we probably should consider.